platform/ios/src/MLNMapView+Metal.mm

```diff
@@ -0,0 +1,227 @@
+        ? [[UIScreen mainScreen] nativeScale]
+        : [[UIScreen mainScreen] scale];
+}
+} // namespace
+
+class MLNMapViewMetalRenderableResource final : public mbgl::mtl::RenderableResource {
+public:
+    MLNMapViewMetalRenderableResource(MLNMapViewMetalImpl& backend_)
+        : backend(backend_),
+          delegate([[MLNMapViewImplDelegate alloc] initWithImpl:&backend]),
+          atLeastiOS_12_2_0([NSProcessInfo.processInfo
+              isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){ 12, 2, 0 }]) {
+    }
+
+    void bind() override {
+        if (!commandQueue) {
+            commandQueue = [mtlView.device newCommandQueue];
+        }
+
+        commandBuffer = [commandQueue commandBuffer];
+        commandBufferPtr = NS::RetainPtr((__bridge MTL::CommandBuffer*)commandBuffer);
+
+        currentDrawable = [mtlView currentDrawable];
+    }
+
+    const mbgl::mtl::RendererBackend& getBackend() const override { return backend; }
+
+    const mbgl::mtl::MTLCommandBufferPtr& getCommandBuffer() const override {
+        return commandBufferPtr;
+    }
+
+    virtual mbgl::mtl::MTLBlitPassDescriptorPtr getUploadPassDescriptor() const override {
+        // Create from render pass descriptor?
+        return NS::TransferPtr(MTL::BlitPassDescriptor::alloc()->init());
+    }
+
+    mbgl::mtl::MTLRenderPassDescriptorPtr getRenderPassDescriptor() const override {
+        return NS::RetainPtr((__bridge MTL::RenderPassDescriptor*)mtlView.currentRenderPassDescriptor);
+    }
+
+    void swap() override {
+        [commandBuffer presentDrawable:currentDrawable];
+        [commandBuffer commit];
+        commandBuffer = nil;
+        commandBufferPtr.reset();
+    }
+
+    mbgl::Size framebufferSize() {
+        assert(mtlView);
+        return { static_cast<uint32_t>(mtlView.drawableSize.width),
+                 static_cast<uint32_t>(mtlView.drawableSize.height) };
+    }
+
+private:
+    MLNMapViewMetalImpl& backend;
+    mbgl::mtl::MTLCommandBufferPtr commandBufferPtr;
+
+public:
+    MLNMapViewImplDelegate* delegate = nil;
+    MTKView *mtlView = nil;
+    id<CAMetalDrawable> currentDrawable;
+    id <MTLCommandBuffer> commandBuffer;
+    id <MTLCommandQueue> commandQueue;
+    const bool atLeastiOS_12_2_0;
+
+    // We count how often the context was activated/deactivated so that we can truly deactivate it
+    // after the activation count drops to 0.
+    NSUInteger activationCount = 0;
+};
+
+MLNMapViewMetalImpl::MLNMapViewMetalImpl(MLNMapView* nativeView_)
+    : MLNMapViewImpl(nativeView_),
+      mbgl::mtl::RendererBackend(mbgl::gfx::ContextMode::Unique),
+      mbgl::gfx::Renderable({ 0, 0 }, std::make_unique<MLNMapViewMetalRenderableResource>(*this)) {
+}
+
+MLNMapViewMetalImpl::~MLNMapViewMetalImpl() {
+    auto& resource = getResource<MLNMapViewMetalRenderableResource>();
+    /*if (resource.context && [[EAGLContext currentContext] isEqual:resource.context]) {
+        [EAGLContext setCurrentContext:nil];
```

Comment:
Is this intentional?

Reply:
Doesn't look familiar, @alexcristici?  I think it's just left over from basing the Metal impl on OpenGL, and can be dropped, as can every trace of global state management other than the Context.

Reply:
Yes, we should remove it. 👍